### PR TITLE
Add password reset interface and API

### LINKED ISSUE #141: Password recovery / reset
It should be possible to automatically recover or reset a forgotten password.

Depends on #140.

### PR DESCRIPTION
**Pull request by cgolubi1**
* Fixes #141

### CODE REVIEW

**Comment by cgolubi1** [comment]
Let's make sure this branch passes CircleCI, then i'll stand up a dev site so people can check this out.  Shadowshade, let me know if you have any other concerns before i do that.

File: src/ui/forgot_password.html [hunk 1]
```diff
@@ -0,0 +1,44 @@
+<!DOCTYPE html>
+<!--
+  This document should be written in polyglot HTML5/XHTML5, and therefore
+  should run when served either as text/html or application/xhtml+xml.
+-->
+<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en">
+<head>
+  <meta charset="utf-8" />
+  <title>Create Account &mdash; Button Men Online</title>
```
Comment (blackshadowshade) [inline src/ui/forgot_password.html]: Did you want to change this title from "Create Account" to something more appropriate?

**Comment by blackshadowshade** [comment]
Apart from that one comment, I see no issues in a cursory read through. I'm happy for this to go to testing.

File: src/ui/forgot_password.html (hunk 1 above)
Comment (cgolubi1) [reply src/ui/forgot_password.html]: Thanks, fixed.

**Comment by cgolubi1** [comment]
Dev site is up --- irilyth, can you attach temporary=34.228.186.22?

**Comment by blackshadowshade** [comment]
So, I've requested a password reset, and see the appropriate entry in the table 'player_reset_verification'.

However, I haven't received an email.

I note also that there was another password reset requested, probably by @cgolubi1. Did you receive an email?

**Comment by cgolubi1** [comment]
Yes, i got my e-mail.

**Comment by cgolubi1** [comment]
Shadowshade, i'm too lazy to sanitize the log message for the website, but of course the issue is that the site doesn't have a DNS name with a PTR record and etc, so gmail blocked your message.  You can see the bounce message in `/var/log/mail.log`

**Comment by blackshadowshade** [comment]
Ah yes, of course.

**Comment by blackshadowshade** [comment]
Is this likely to work on staging, or will we have the same problem there?

**Comment by irilyth** [comment]
> Dev site is up --- irilyth, can you attach temporary=34.228.186.22?

Done!

**Comment by cgolubi1** [comment]
Shadowshade: i requested a password reset for my own account and forwarded you the e-mail.  You should be able to use it.

**Comment by blackshadowshade** [comment]
The email looks okay, the password reset link looks okay, I was able to reset the password and then log in with the new password.

This looks good to go from my side. I'll merge this.

**Comment by blackshadowshade** [comment]
Just a note that the password reset link doesn't get disabled after it's been used, that may be desirable for the future.

**Comment by AdmiralJota** [comment]
Does it expire?

**Comment by blackshadowshade** [comment]
No, it doesn't expire either.

**Comment by AdmiralJota** [comment]
Is that going to be safe? I'm concerned that several months (or years) after someone requests a reset, if they don't delete the email, anyone who manages to get a hold of it would have permanent control over the user's account, being able to re-set their password repeatedly at will.

**Comment by blackshadowshade** [comment]
I think I agree, the reset should probably be valid for a short period, maybe an hour or so, and it should expire when used. I'll open a new issue for that.

**Pull request merged**